client/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -128,12 +128,7 @@ final case class Client[F[_]](
   def streaming[A](req: F[Request[F]])(f: Response[F] => Stream[F, A]): Stream[F, A] =
     Stream.eval(req).flatMap(streaming(_)(f))
 
-  /**
-    * Submits a request and decodes the response on success.  On failure, the
-    * status code is returned.  The underlying HTTP connection is closed at the
-    * completion of the decoding.
-    */
-  def expect[A](req: Request[F])(implicit d: EntityDecoder[F, A]): F[A] = {
+  def expectOr[A](req: Request[F])(onError: Response[F] => F[Throwable])(implicit d: EntityDecoder[F, A]): F[A] = {
```

Comment:
I think it can be generalized to
```scala
def expectOr[A](req: Request[F])(onError: Response[F] => F[A])(implicit d: EntityDecoder[F, A]): F[A]
```

Reply:
The downside then is it's a bit more cumbersome to provide a generic onError handler for all `A`s. Especially if we want make it part of the client conf. But I can see benefits of being able to directly recover from response .

Reply:
I don't think this is a generalization when you might possibly not want to return an `A`.

It's much easier to handle this outside the function (via handleError) than not. I think I prefer the way it is.

Reply:
If you don't want to return an `A`, you can just `F.raiseError(whatever)`.

Reply:
Returing F[A] is more flexible but the API  gives the impression that the intention here is a recovery. User can always do a recovery with F after, so making it raising an error on non-200 status seems a good design to me.